iModelCore/ECDb/Tests/NonPublished/ECSqlAliasResolutionTests.cpp

```diff
@@ -10,29 +10,64 @@ struct ECSqlAliasResolutionTestFixture : ECDbTestFixture {};
-    ECSQL("SELECT (SELECT ECInstanceId FROM Bis.Model m WHERE m.ECInstanceId = e.Model.Id) FROM Bis.Element e");
-    ECSQL("SELECT (SELECT ECInstanceId FROM (SELECT m.ECInstanceId FROM Bis.Model m WHERE m.ECInstanceId = e.Model.Id)) FROM Bis.Element e");
-    ECSQL("SELECT c.ECInstanceId, m.ECInstanceId FROM bis.SpatialCategory c JOIN bis.Model m ON m.ECInstanceId IN (SELECT e.Model.Id FROM bis.GeometricElement3d e WHERE e.Category.Id = c.ECInstanceId)");
-    ECSQL("SELECT (SELECT ECInstanceId FROM (SELECT C.ECInstanceId FROM Meta.ECClassDef C WHERE C.ECInstanceId = E.ECClassId)) FROM Bis.Element E")
-    ECSQL("SELECT (SELECT ECInstanceId FROM (SELECT ECInstanceId FROM Meta.ECClassDef WHERE ECInstanceId = E.ECClassId)) FROM Bis.Element E")
+    ASSERT_ECSQL("SELECT 1 FROM meta.ECClassDef, (SELECT ECInstanceId FROM meta.ECPropertyDef) S WHERE ECClassDef.ECInstanceId=S.ECInstanceId");
+    ASSERT_ECSQL("SELECT e.* FROM Bis.Element e");
+    ASSERT_ECSQL("SELECT e.* FROM Bis.Element e JOIN Bis.Model m ON m.ECInstanceId = e.Model.Id");
+    ASSERT_ECSQL("SELECT e.* FROM Bis.Element e WHERE e.Model.Id IN (SELECT m.ECInstanceId FROM Bis.Model m WHERE m.ECInstanceId = e.Model.Id)");
+    ASSERT_ECSQL("SELECT (SELECT ECInstanceId FROM Bis.Model m WHERE m.ECInstanceId = e.Model.Id) FROM Bis.Element e");
+    ASSERT_ECSQL("SELECT (SELECT ECInstanceId FROM (SELECT m.ECInstanceId FROM Bis.Model m WHERE m.ECInstanceId = e.Model.Id)) FROM Bis.Element e");
+    ASSERT_ECSQL("SELECT c.ECInstanceId, m.ECInstanceId FROM bis.SpatialCategory c JOIN bis.Model m ON m.ECInstanceId IN (SELECT e.Model.Id FROM bis.GeometricElement3d e WHERE e.Category.Id = c.ECInstanceId)");
+    ASSERT_ECSQL("SELECT (SELECT ECInstanceId FROM (SELECT C.ECInstanceId FROM Meta.ECClassDef C WHERE C.ECInstanceId = E.ECClassId)) FROM Bis.Element E")
+    ASSERT_ECSQL("SELECT (SELECT ECInstanceId FROM (SELECT ECInstanceId FROM Meta.ECClassDef WHERE ECInstanceId = E.ECClassId)) FROM Bis.Element E")
+}
+
+//---------------------------------------------------------------------------------------
+// @bsiclass
+//+---------------+---------------+---------------+---------------+---------------+------
+TEST_F(ECSqlAliasResolutionTestFixture, ColumnAliasesWithClauses) {
+    ASSERT_OPEN_TEST_BIM;
+    EXPECT_ECSQL("select * from (select 'Test' as literalAlias, COUNT(*) from meta.ECClassDef group by literalAlias)")
+
+    EXPECT_ECSQL("select ECInstanceId as idAlias from meta.ECClassDef where idAlias < 30")
+    EXPECT_ECSQL("select 'Test' as literalAlias from meta.ECClassDef where literalAlias < 30")
+
+    EXPECT_ECSQL("select ECInstanceId as idAlias from meta.ECClassDef group by idAlias")
+    EXPECT_ECSQL("select 'Test' as literalAlias from meta.ECClassDef group by literalAlias")
+
+    EXPECT_ECSQL("select ECInstanceId as idAlias from meta.ECClassDef order by idAlias desc")
```

Comment:
The test could be
1. `SELECT COUNT(*) cnt FROM test.Foo GROUP BY ECClassId HAVING cnt>10`
2. `SELECT (SELECT 1) a FROM test.Foo WHERE a <> 1`
3. `SELECT p.ECInstanceId=k.s FROM test.Foo, (SELECT 1 s) k